src/app/lightning/update-ln-payments.ts

```diff
@@ -0,0 +1,110 @@
+import { LndService } from "@services/lnd"
+import { baseLogger } from "@services/logger"
+import { LnPaymentsRepository } from "@services/mongoose/ln-payments"
+
+export const updateLnPayments = async (): Promise<true | ApplicationError> => {
+  let processedLnPaymentsHashes: PaymentHash[] | ApplicationError = []
+
+  const incompleteLnPayments = await LnPaymentsRepository().listIncomplete()
+  if (incompleteLnPayments instanceof Error) return incompleteLnPayments
+
+  const lndService = LndService()
+  if (lndService instanceof Error) return lndService
+  const pubkeys = lndService.listActivePubkeys()
+
+  let lastSeenCursorByPubkey: ListSettledAndFailedLnPaymentsByPubkeyArgs = pubkeys.map(
+    (pubkey) => ({
+      settledAfter: undefined,
+      failedAfter: undefined,
+      pubkey,
+    }),
+  )
+
+  while (
+    processedLnPaymentsHashes.length < incompleteLnPayments.length &&
+    lastSeenCursorByPubkey.length
+  ) {
+    ;({ lastSeenCursorByPubkey, processedLnPaymentsHashes } =
+      await fetchAndUpdatePayments({
+        incompleteLnPayments,
+        lastSeenCursorByPubkey,
+        lndService,
+      }))
+  }
+
+  return true
+}
+
+const fetchAndUpdatePayments = async ({
+  incompleteLnPayments,
+  lastSeenCursorByPubkey,
+  lndService,
+}: {
+  incompleteLnPayments: PersistedLnPaymentLookup[]
+  lastSeenCursorByPubkey: ListSettledAndFailedLnPaymentsByPubkeyArgs
+  lndService: ILightningService
+}) => {
+  // Fetch from Lightning service
+  const results = await lndService.listSettledAndFailedPaymentsByPubkey(
+    lastSeenCursorByPubkey,
+  )
+
+  // Update cursors & drop completed pubkeys
```

Comment:
I feel the logic of wrapping around pubkey to be convoluted:
- if we get multiple node to fetch over, it will be just few of them, maybe 2 or 3.
- I would favor simplifying the logic instead of trying to parallelize this code (which I think this is what we're trying to achieve here but maybe I'm mistaken): 

this is a cronjob so it's not critical to get high throughput, and iterating slowly over the payment is actually beneficial to not overload the different lnds.

so maybe a more simple approach in the like of

```
for await (const pubkey of pubkeys) {
 fetchAndUpdatePayments({pubkey, ...})
}
```

might make it easier to understand what is going on.

not a strong opinion on this as I'm trying to understand the PR still, but this is my first reaction.

Reply:
The idea here isn't so much to parallelize as it is to enforce that we look across lnds before we go down each one (breadth-first search).

Say for example these are our 3 lnds and we have relevant payments in the boxes marked `x` that we need to update from. Each box represents one page in the paginated search.

| lnd1 | lnd2 | lnd3 |
|------|------|------|
|   x  |   x  |   x  |
|   -  |   -  |   -  |
|   x  |   -  |   -  |
|   -  |   -  |   -  |
|   -  |   -  |   -  |
|   -  |   -  |   -  |
|   -  |   -  |   -  |
|   -  |   -  |   -  |

If we were to do one lnd at a time, then we would have to query down the entire `lnd1` first, and then `lnd2`, and we won't complete until we get to the first page of the `lnd3`.

What I'm trying to do is query across all lnds first so that our queries would finish on the 3rd page of `lnd1` and we would only need 3 passes instead of potentially much more if we're deleting payments every 2 weeks but running the cron task each day (or multiple times per day).

The intention is to try to avoid a significant number of unnecessary queries to the lnds in an environment where we have multiple lnds.

Reply:
I understand the idea, I'm not sure this is the right tradeoff. 

doing it by lnd would be:

1-0, 1-1, 1-2, 2-0, 3-0

doing it with the cursors the way it's implemented is:

1-0, 2-0, 3-0, 1-1, 1-2

The reason why we would benefits from the added algorithm complexity is if this was on a critical path for latency for a user request. but because this run as a cronjob, there is no reason to optimize for latency (assumption is we want to optimize for the fastest number of payment in the shorter time and have the potential gap later in the process?).

I'm also not saying we have to rewrite it. but I just think as a result the code is not easy to understand.

Reply:
Agreed this makes the code more complex.

Maybe a good tradeoff could be to leave it like this for now until we implement the `deletePayments` logic, and then simplify at that point when we have much lighter lnd payment histories to parse.